File: sdk/src/main/java/com/meniga/sdk/models/sync/MenigaSync.java
package com.meniga.sdk.models.sync;

import android.os.Handler;
import android.os.Parcel;
import android.os.Parcelable;

import org.joda.time.DateTime;

import java.io.Serializable;
import java.util.List;

import com.meniga.sdk.MenigaSDK;
import com.meniga.sdk.helpers.Interceptor;
import com.meniga.sdk.helpers.Result;
import com.meniga.sdk.models.sync.operators.MenigaSyncOperations;

/**
 * Represents the status of a sync procedure.
 * <p>
 * Copyright 2017 Meniga Iceland Inc.
 */
public class MenigaSync implements Serializable, Parcelable, Cloneable {

	protected static MenigaSyncOperations apiOperator;

	protected long syncHistoryId;
	protected boolean isSyncDone;
	protected DateTime syncSessionStartTime;
	protected List<RealmSyncResponse> realmSyncResponses;

	protected MenigaSync() {
	}

	protected MenigaSync(Parcel in) {
		this.syncHistoryId = in.readLong();
		this.isSyncDone = in.readByte() != 0;
		this.syncSessionStartTime = (DateTime) in.readSerializable();
		this.realmSyncResponses = in.createTypedArrayList(RealmSyncResponse.CREATOR);
	}

	@Override
	public int describeContents() {
		return 0;
	}

	@Override
	public void writeToParcel(Parcel dest, int flags) {
		dest.writeLong(this.syncHistoryId);
		dest.writeByte(this.isSyncDone ? (byte) 1 : (byte) 0);
		dest.writeSerializable(this.syncSessionStartTime);
		dest.writeTypedList(this.realmSyncResponses);
	}

	public static final Creator<MenigaSync> CREATOR = new Creator<MenigaSync>() {
		@Override
		public MenigaSync createFromParcel(Parcel source) {
			return new MenigaSync(source);
		}

		@Override
		public MenigaSync[] newArray(int size) {
			return new MenigaSync[size];
		}
	};

	@Override
	protected MenigaSync clone() throws CloneNotSupportedException {
		return (MenigaSync) super.clone();
	}

	/**
	 * Sets the api operator for doing api calls
	 *
	 * @param operator An object that implements the MenigaSyncOperations interface for carrying out api operations on this class.
	 */
	public static void setOperator(MenigaSyncOperations operator) {
		MenigaSync.apiOperator = operator;
	}

	private boolean hasNewData() {
		if (realmSyncResponses == null || realmSyncResponses.size() == 0) {
			return false;
		}
		for (RealmSyncResponse resp : realmSyncResponses) {
			for (AccountSyncStatus stat : resp.getAccountSyncStatuses()) {
				if (stat.getTransactionsProcessed() > 0) {
					return true;
				}
			}
		}
		return false;
	}

	/**
	 * @return Number of new transactions that came into the system as a result of the sync.
	 */
	public int getNumNewTransactions() {
		if (realmSyncResponses == null || realmSyncResponses.size() == 0) {
			return 0;
		}
		int cnt = 0;
		for (RealmSyncResponse resp : realmSyncResponses) {
			for (AccountSyncStatus stat : resp.getAccountSyncStatuses()) {
				cnt += stat.getTransactionsProcessed();
			}
		}
		return cnt;
	}

	public long getSyncHistoryId() {
		return syncHistoryId;
	}

	public DateTime getSyncSessionStartTime() {
		return syncSessionStartTime;
	}

	public List<RealmSyncResponse> getRealmSyncResponses() {
		return realmSyncResponses;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}

		MenigaSync that = (MenigaSync) o;

		if (syncHistoryId != that.syncHistoryId) {
			return false;
		}
		if (isSyncDone != that.isSyncDone) {
			return false;
		}
		if (syncSessionStartTime != null ? !syncSessionStartTime.equals(that.syncSessionStartTime) : that.syncSessionStartTime != null) {
			return false;
		}
		return realmSyncResponses != null ? realmSyncResponses.equals(that.realmSyncResponses) : that.realmSyncResponses == null;
	}

	@Override
	public int hashCode() {
		int result = (int) (syncHistoryId ^ (syncHistoryId >>> 32));
		result = 31 * result + (isSyncDone ? 1 : 0);
		result = 31 * result + (syncSessionStartTime != null ? syncSessionStartTime.hashCode() : 0);
		result = 31 * result + (realmSyncResponses != null ? realmSyncResponses.hashCode() : 0);
		return result;
	}

	/**
	 * A class used as a callback. It will be called when the sync precedure is done or if it fails or times out.
	 * <p>
	 * Copyright 2017 Meniga Iceland Inc.
	 */
	public static abstract class PostSyncCallback {
		protected boolean hasNewData;
		protected Long timeStamp;
		protected int numNewTransactions;
		protected MenigaSync sync;

		public boolean getHasNewData() {
			return this.hasNewData;
		}

		public abstract void onFailure(Exception exception);

		public abstract void onFinished();

		public abstract void onTimedOut();

		/**
		 * @param millis Sets the start time stamp.
		 */
		public void setStartTimeStamp(long millis) {
			this.timeStamp = millis;
		}

		/**
		 * @return The start time stamp.
		 */
		public Long getStartTimeStamp() {
			return this.timeStamp;
		}

		/**
		 * @return Number of new transactions that came into the system as a result of the sync.
		 */
		public int getNumNewTransactions() {
			return this.numNewTransactions;
		}
	}

	/*
	--- API calls below ---
	 */

	private void checkSync(final PostSyncCallback postSync, final long timeout, final long interval) {
		if (postSync.getStartTimeStamp() == null) {
			postSync.setStartTimeStamp(DateTime.now().getMillis());
		}
		Handler handler = new Handler();
		handler.postDelayed(new Runnable() {
			@Override
			public void run() {
				Result<MenigaSyncStatus> task = isSyncDone();
				MenigaSDK.getMenigaSettings().getTaskAdapter().intercept(task, new Interceptor<MenigaSyncStatus>() {
					@Override
					public void onFinished(MenigaSyncStatus result, boolean failed) {

						long delta = DateTime.now().getMillis() - postSync.getStartTimeStamp();
						if (failed || result == null) {
							postSync.onFailure(new Exception("Failure syncing"));
						} else if (result.getHasCompletedSyncSession()) {
							Result<MenigaSync> subTask = MenigaSync.fetch(syncHistoryId);
							MenigaSDK.getMenigaSettings().getTaskAdapter().intercept(subTask, new Interceptor<MenigaSync>() {
								@Override
								public void onFinished(MenigaSync result, boolean failed) {
									if (failed || result == null) {
										postSync.onFailure(new Exception("Failure syncing"));
									}
									postSync.hasNewData = result != null && result.hasNewData();
									postSync.numNewTransactions = result != null ? result.getNumNewTransactions() : 0;
									postSync.sync = result;
									postSync.onFinished();
								}
							});
						} else if (delta > timeout) {
							postSync.onTimedOut();
						} else {
							checkSync(postSync, timeout, interval);
						}
					}
				});
			}
		}, interval);
	}

	/**
	 * Gets a sync object that was created by MenigaSync.start
	 *
	 * @param syncHistoryId The id of the sync object
	 * @return The sync object
	 */
	public static Result<MenigaSync> fetch(long syncHistoryId) {
		return MenigaSync.apiOperator.getSync(syncHistoryId);
	}

	/**
	 * Starts the accounts synchronization process and also returns a MenigaSync object
	 * that contains further details.
	 *
	 * @param timeout Timeout for the sync procedure, the actuall response callback might take longer
	 * @return A new sync object created by starting the sync procedure
	 */
	public static Result<MenigaSync> start(long timeout) {
		return MenigaSync.start(timeout, 1000, null);
	}

	/**
	 * Starts the accounts synchronization process and also returns a MenigaSync object
	 * that contains further details.
	 *
	 * @param timeout  The amount of time the background process should try to check the syn result before terminating.
	 * @param interval The amount of time between checks for sync completion in the background
	 * @return A new sync object.
	 */
	public static Result<MenigaSync> start(final long timeout, final long interval, final Interceptor<MenigaSync> onDone) {
		Result<MenigaSync> task = MenigaSync.apiOperator.startSync(timeout);
		return MenigaSDK.getMenigaSettings().getTaskAdapter().intercept(task, new Interceptor<MenigaSync>() {
			@Override
			public void onFinished(final MenigaSync result, boolean failed) {
				if (result == null) {
					if (onDone != null) {
						onDone.onFinished(null, true);
					}
					return;
				}
				result.checkSync(new PostSyncCallback() {
					@Override
					public void onFailure(Exception excp) {
						if (onDone != null) {
							onDone.onFinished(null, true);
						}
					}

					@Override
					public void onFinished() {
						if (onDone != null) {
							onDone.onFinished(sync, false);
						}
					}

					@Override
					public void onTimedOut() {
						if (onDone != null) {
							onDone.onFinished(null, true);
						}
					}
				}, timeout, interval);
			}
		});
	}

	/**
	 * Checks to see if the synchronization operation has finished
	 *
	 * @return A Task containing a boolean indicating if the synchronization is done or not
	 */
	public Result<MenigaSyncStatus> isSyncDone() {
		return MenigaSync.apiOperator.getSyncStatus();
	}
}

File: sdk/src/main/java/com/meniga/sdk/models/sync/MenigaSync.kt
/*
 * Copyright 2017-2018 Meniga Iceland Inc.
 */
package com.meniga.sdk.models.sync

import android.os.Handler
import android.os.Parcelable
import com.meniga.sdk.MenigaSDK
import com.meniga.sdk.helpers.Interceptor
import com.meniga.sdk.helpers.Result
import com.meniga.sdk.models.sync.operators.MenigaSyncOperations
import kotlinx.android.parcel.Parcelize
import org.joda.time.DateTime
import java.io.Serializable

/**
 * Represents the status of a sync procedure.
 */
@Parcelize
@SuppressWarnings("ParcelCreator")
data class MenigaSync(
        val syncHistoryId: Long = 0,
        val isSyncDone: Boolean = false,
        val syncSessionStartTime: DateTime? = null,
        val realmSyncResponses: List<RealmSyncResponse>? = null
): Serializable, Parcelable, Cloneable {

    /**
     * @return Number of new transactions that came into the system as a result of the sync.
     */
    val numNewTransactions: Int
        get() = realmSyncResponses
                ?.flatMap { it.getAccountSyncStatuses() }
                ?.sumBy { it.getTransactionsProcessed() }
                ?: 0

    @Throws(CloneNotSupportedException::class)
    override fun clone(): MenigaSync {
        return super.clone() as MenigaSync
    }

    /**
     * A class used as a callback. It will be called when the sync procedure is done or if it fails or times out.
     */
    abstract class PostSyncCallback {
        var hasNewData: Boolean = false
        /**
         * @return The start time stamp.
         */
        var startTimeStamp: Long? = null

        /**
         * @return Number of new transactions that came into the system as a result of the sync.
         */
        var numNewTransactions: Int = 0
        var sync: MenigaSync? = null

        abstract fun onFailure(exception: Exception)

        abstract fun onFinished()

        abstract fun onTimedOut()
    }

    /*
	--- API calls below ---
	 */

    private fun checkSync(postSync: PostSyncCallback, timeout: Long, interval: Long) {
        if (postSync.startTimeStamp == null) {
            postSync.startTimeStamp = DateTime.now().millis
        }
        val handler = Handler()
        handler.postDelayed({
            val task = getSyncStatus()
            MenigaSDK.getMenigaSettings().taskAdapter.intercept(task, object : Interceptor<MenigaSyncStatus>() {
                override fun onFinished(result: MenigaSyncStatus?, failed: Boolean) {

                    val delta = DateTime.now().millis - postSync.startTimeStamp!!
                    if (failed || result == null) {
                        postSync.onFailure(Exception("Failure syncing"))
                    } else if (result.getHasCompletedSyncSession()) {
                        val subTask = MenigaSync.fetch(syncHistoryId)
                        MenigaSDK.getMenigaSettings().taskAdapter.intercept(subTask, object : Interceptor<MenigaSync>() {
                            override fun onFinished(result: MenigaSync?, failed: Boolean) {
                                if (failed || result == null) {
                                    postSync.onFailure(Exception("Failure syncing"))
                                }
                                postSync.hasNewData = result != null && result.hasNewData()
                                postSync.numNewTransactions = result?.numNewTransactions ?: 0
                                postSync.sync = result
                                postSync.onFinished()
                            }
                        })
                    } else if (delta > timeout) {
                        postSync.onTimedOut()
                    } else {
                        checkSync(postSync, timeout, interval)
                    }
                }
            })
        }, interval)
    }

    private fun hasNewData(): Boolean = realmSyncResponses
            ?.flatMap { it.getAccountSyncStatuses() }
            ?.firstOrNull { it.getTransactionsProcessed() > 0 } != null

    /**
     * Checks to see if the synchronization operation has finished
     *
     * @return A Task containing a boolean indicating if the synchronization is done or not
     */
    @Deprecated("Use static getSyncStatus instead", replaceWith = ReplaceWith("getSyncStatus"))
    fun isSyncDone(): Result<MenigaSyncStatus> = apiOperator.syncStatus

    companion object {
        private const val SYNC_CHECK_INTERVAL_MILLISECONDS: Long = 1000

        private lateinit var apiOperator: MenigaSyncOperations

        /**
         * Sets the api operator for doing api calls
         *
         * @param operator An object that implements the MenigaSyncOperations interface for carrying out api operations on this class.
         */
        @JvmStatic
        fun setOperator(operator: MenigaSyncOperations) {
            MenigaSync.apiOperator = operator
        }

        /**
         * Gets a sync object that was created by [start]
         *
         * @param syncHistoryId The id of the sync object
         * @return The sync object
         */
        @JvmStatic
        fun fetch(syncHistoryId: Long): Result<MenigaSync> {
            return MenigaSync.apiOperator.getSync(syncHistoryId)
        }

        /**
         * Use [syncRealms] instead.
         */
        @Deprecated("Use syncRealms instead")
        @JvmStatic
        fun start(timeout: Long): Result<MenigaSync> {
            return launchSync(null, timeout, null)
        }

        /**
         * Use [syncRealms] instead.
         */
        @Deprecated("Use syncRealms instead")
        @JvmStatic
        fun start(timeout: Long, interval: Long, onDone: Interceptor<MenigaSync>): Result<MenigaSync> {
            return launchSync(null, timeout, onDone)
        }

        /**
         * Starts the accounts synchronization process and also returns a MenigaSync object
         * that contains further details.
         *
         * @param timeout Timeout for the sync procedure.
         * @return A new sync object created by starting the sync procedure
         */
        @JvmStatic
        fun syncRealms(timeout: Long): Result<MenigaSync> {
            return launchSync(null, timeout, null)
        }

        /**
         * Starts the accounts synchronization process and also returns a MenigaSync object
         * that contains further details.
         *
         * @param timeout  The amount of time the background process should try to check if the sync has completed before terminating.
         * @return A new sync object.
         */
        @JvmStatic
        fun syncRealms(timeout: Long, onDone: Interceptor<MenigaSync>? = null): Result<MenigaSync> {
            return launchSync(null, timeout, onDone)
        }

        /**
         * Starts the accounts synchronization process for a specific realm and also returns a MenigaSync object
         * that contains further details.
         *
         * @param realmUserId  The id of the realm account user - a realm is a "department" in e.g. a bank. Most organizations (most often banks) have only one but some
         * large organizations can have many realms (e.g. insurance, banking, credit cards and so on). This id identifies the user's realm user account.
         * @param timeout  The amount of time the background process should try to check if the sync has completed before terminating.
         * @return A new sync object.
         */
        @JvmStatic
        fun syncRealm(realmUserId: Long, timeout: Long, onDone: Interceptor<MenigaSync>? = null): Result<MenigaSync> {
            return launchSync(realmUserId, timeout, onDone)
        }

        private fun launchSync(realmUserId: Long?, timeout: Long, onDone: Interceptor<MenigaSync>?): Result<MenigaSync> {
            val task = getSyncStatus().task.continueWithTask { task ->
                if (task.isFaulted || task.result.hasCompletedSyncSession) {
                    apiOperator.startSync(realmUserId, timeout).task
                } else {
                    fetch(task.result.getSynchronizationStatus().syncHistoryId).task
                }
            }
            return MenigaSDK.getMenigaSettings().taskAdapter.intercept(task, object : Interceptor<MenigaSync>() {
                override fun onFinished(result: MenigaSync?, failed: Boolean) {
                    if (result == null) {
                        onDone?.onFinished(null, true)
                        return
                    }
                    result.checkSync(object : PostSyncCallback() {
                        override fun onFailure(exception: Exception) {
                            onDone?.onFinished(null, true)
                        }

                        override fun onFinished() {
                            onDone?.onFinished(sync, false)
                        }

                        override fun onTimedOut() {
                            onDone?.onFinished(null, true)
                        }
                    }, timeout, SYNC_CHECK_INTERVAL_MILLISECONDS)
                }
            })
        }

        /**
         * Checks to see the status of all synchronization operations on the back end
         *
         * @return A Task indicating if the synchronization is done or not
         */
        @JvmStatic
        fun getSyncStatus(): Result<MenigaSyncStatus> = apiOperator.syncStatus
    }
}
